presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlock.java

```diff
@@ -464,4 +465,74 @@ public DictionaryBlock compact()
             return this;
         }
     }
+
+    @Override
+    public byte getByteUnchecked(int internalPosition)
+    {
+        assert internalPositionInRange(internalPosition, getOffsetBase(), getPositionCount());
+        return dictionary.getByte(ids[internalPosition]);
```

Comment:
I've decided that the scope of `UncheckedBlock` should only include the boundary checks on the current block, and never the nested block.  Otherwise the behavior and implementation become inconsistent and difficult to reason about.  `DictionaryBlock` is a good example--if I were to use the internal `getByteUnchecked`, it's not clear how I would also expressed the internal block's offset.  I could do this:

`dictionary.getByte(ids[internalPosition] + dictionary.getOffsetBase());`

But it's more efficient to allow one to unpeel the dictionary and iterate over it manually.  It's also easier to reason about IMO: unchecked blocks give you the block's internal offset, but for nested structures, if you want to avoid the performance penalty imposed by the nesting, you need to unpeel them using the existing utilities already available (`ColumnarX` classes, `BlockFlattener`).  Otherwise there will always be gotchas (how would this work for doubly nested dictionary, for example).

Reply:
Makes sense. Is there a place to write this down?

Reply:
I'll add this javadoc to the existing comment in `UncheckedBlock`:

```java
 * For nested structures, such as dictionaries and RLEs, the indexing is unchecked with respect to the top
 * level block, but not with respect to the inner blocks.  If, for performance reasons, it is desired to
 * use unchecked indexing, you may unpeel the blocks using existing utilities such as {@link ColumnarArray},
 * {@link ColumnarMap}, {@link ColumnarRow} and {@link BlockFlattener}.
```